storage/src/adb/keyless.rs

```diff
@@ -112,163 +83,87 @@ pub struct Keyless<E: Storage + Clock + Metrics, V: Codec, H: CHasher> {
-        locations: &mut FJournal<E, u32>,
-        log: &mut VJournal<E, Operation<V>>,
-        last_log_op: Operation<V>,
-        op_count: Location,
-        last_offset: u32,
-        log_items_per_section: u64,
-    ) -> Result<Location, Error> {
-        let mut first_uncommitted: Option<(Location, u32)> = None;
-        let mut op_index = op_count - 1;
-        let mut op = last_log_op;
-        let mut offset = last_offset;
-        let oldest_retained_loc = locations
-            .oldest_retained_pos()
-            .await?
-            .expect("location should be nonempty");
-
-        // Walk backwards through the log until we find the last commit point.
         loop {
+            let op = log.read(loc).await?;
             match op {
-                Operation::Commit(_) => {
-                    break;
-                }
+                Operation::Commit(_) => break,
                 Operation::Append(_) => {
-                    // Track the earliest uncommitted append (index, offset) encountered while
-                    // walking backwards. If none is found before we hit a Commit, there is
-                    // nothing to rewind.
-                    first_uncommitted = Some((op_index, offset));
+                    first_uncommitted = Some(loc);
                 }
             }
-            if op_index == oldest_retained_loc {
-                assert_eq!(op_index, 0, "no commit operation found");
+            if loc == oldest_retained_loc {
                 break;
             }
-            op_index -= 1;
-            offset = locations.read(*op_index).await?;
-            let section = *op_index / log_items_per_section;
-            op = log.get(section, offset).await?;
+            loc -= 1;
         }
 
-        // If there are no uncommitted operations, exit early.
-        let Some((rewind_size, rewind_offset)) = first_uncommitted else {
-            return Ok(op_index + 1);
-        };
+        // Rewind operations after the last commit
+        if let Some(rewind_loc) = first_uncommitted {
+            let ops_to_rewind = log_size - rewind_loc;
+            warn!(ops_to_rewind, ?rewind_loc, "rewinding log to last commit");
+            log.rewind(rewind_loc).await?;
+            log.sync().await?;
+            Ok(rewind_loc)
+        } else {
+            Ok(log_size)
+        }
+    }
 
-        // Rewind the log and MMR to the last commit point.
-        let ops_to_rewind = (*op_count - *rewind_size) as usize;
-        warn!(ops_to_rewind, ?rewind_size, "rewinding log to last commit");
-        locations.rewind(*rewind_size).await?;
-        locations.sync().await?;
-        mmr.pop(ops_to_rewind).await?; // sync is handled by pop
-        let section = *rewind_size / log_items_per_section;
-        log.rewind_to_offset(section, rewind_offset).await?;
-        log.sync(section).await?;
-
-        Ok(rewind_size)
+    /// Align `mmr` with `log` by either popping excess operations or replaying missing ones.
```

Comment:
My (pending) storage-code-sharing PR has a shared implementation we could use here, just in case it gets merged first.